tests/conftest.py

```diff
@@ -67,16 +71,85 @@ def events_log():
 
 
 @pytest.fixture
-def late_events(messages, events_log):
+def events_log(messages, events_log_init):
     late_events_messages = [message.get("timestamp") for message in messages]
-    events_log["late_events"]["seconds"] = late_events_messages
-    events_log["late_events"]["minutes"] = late_events_messages
-    events_log["late_events"]["hours"] = late_events_messages
-    events_log["late_events"]["days"] = late_events_messages
-    return events_log
+    events_log_init["late_events"]["seconds"] = late_events_messages
+    events_log_init["late_events"]["minutes"] = late_events_messages
+    events_log_init["late_events"]["hours"] = late_events_messages
+    events_log_init["late_events"]["days"] = late_events_messages
+    return events_log_init
 
 
 @pytest.fixture
-def past_events(messages, events_log):
-    events_log["past_events"] = messages
-    return events_log
+def past_events(messages, events_log_init):
+    events_log_init["past_events"] = messages
+    return events_log_init
+
+
+@pytest.fixture
+def pubsub():
+    return get_topic(pubsub=False)
+
+
+@pytest.fixture
+def late_events(faker_vehicle, events_log, pubsub, tollways):
+    late_event = LateEventProcessor(
+        events_log=events_log,
+        publisher=pubsub[0],
+        topic_path=pubsub[1],
+        fake=faker_vehicle["faker"],
+        tollways=tollways,
+    )
+    late_events = {}
+    for time_unit in TIME_UNIT.keys():
+        late_event.time_unit = time_unit
+        late_events[time_unit] = late_event.create_event()
+    return late_events
+
+
+@pytest.fixture
+def get_microseconds():
+    def _get_microseconds(ts: str):
+        return datetime.strptime(ts, TIMESTAMP_FORMAT).microsecond
+
+    return _get_microseconds
+
+
+@pytest.fixture
+def generate_event_pair(get_microseconds, late_events, events_log):
+    def _generate_event_pair(time_unit):
+        late_event_timestamp = late_events[time_unit]["timestamp"]
+        late_event_microseconds = get_microseconds(late_event_timestamp)
+
+        for original_event_timestamp in events_log["late_events"][time_unit]:
+            original_event_microseconds = get_microseconds(original_event_timestamp)
+            # Compare the microseconds part of the timestamps
+            if late_event_microseconds == original_event_microseconds:
+                return {
+                    "original": datetime.strptime(original_event_timestamp, TIMESTAMP_FORMAT),
+                    "late": datetime.strptime(late_event_timestamp, TIMESTAMP_FORMAT),
+                }
+        return None
```

Comment:
would be more consistent to always return a dict?

Reply:
good call